core/state/snapshot/disklayer.go

```diff
@@ -84,14 +82,13 @@ func (dl *diskLayer) Account(hash common.Hash) (*types.SlimAccount, error) {
 // AccountRLP directly retrieves the account RLP associated with a particular
 // hash in the snapshot slim data format.
 func (dl *diskLayer) AccountRLP(hash common.Hash) ([]byte, error) {
-	dl.lock.RLock()
-	defer dl.lock.RUnlock()
-
 	// If the layer was flattened into, consider it invalid (any live reference to
 	// the original should be marked as unusable).
-	if dl.stale {
+	if dl.stale.Load() {
 		return nil, ErrSnapshotStale
 	}
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
```

Comment:
Unfortunately, this changes the semantics in a suble but breaking way. Before this change, at line `97`, we can be certain that the layer is not stale, since we obtain the readlock, and nobody can set it to stale without having the writelock. 
After the change, you check atomically, and _then_ obtain the readlock. But before the readlock is obtained, someoen might have made this layer stale. 

So, changing the behaviour here is non-trivial, since it's not just about the boolean itself, but what the boolean signifies. All in all, let's just leave it as it is, it's not worth the risk of changing it just to be more "idiomatic" or whatever.